src/Shared/Dictionary/AdaptiveCapacityDictionary.cs

```diff
@@ -0,0 +1,752 @@
+
+        /// <inheritdoc />
+        public ICollection<TKey> Keys
+        {
+            get
+            {
+                if (_dictionaryStorage != null)
+                {
+                    return _dictionaryStorage.Keys;
+                }
+
+                var array = _arrayStorage;
+                var keys = new TKey[_count];
+                for (var i = 0; i < keys.Length; i++)
+                {
+                    keys[i] = array[i].Key;
+                }
+
+                return keys;
+            }
+        }
+
+        IEnumerable<TKey> IReadOnlyDictionary<TKey, TValue>.Keys => Keys;
+
+        /// <inheritdoc />
+        public ICollection<TValue> Values
+        {
+            get
+            {
+                if (_dictionaryStorage != null)
+                {
+                    return _dictionaryStorage.Values;
+                }
+
+                var array = _arrayStorage;
+                var values = new TValue[_count];
+                for (var i = 0; i < values.Length; i++)
+                {
+                    values[i] = array[i].Value;
+                }
+
+                return values;
+            }
+        }
+
+        IEnumerable<TValue> IReadOnlyDictionary<TKey, TValue>.Values => Values;
+
+        /// <inheritdoc />
+        void ICollection<KeyValuePair<TKey, TValue>>.Add(KeyValuePair<TKey, TValue> item)
+        {
+            if (_dictionaryStorage != null)
+            {
+                ((ICollection<KeyValuePair<TKey, TValue>>)_dictionaryStorage).Add(item);
+                return;
+            }
+
+            Add(item.Key, item.Value);
+        }
+
+        /// <inheritdoc />
+        public void Add(TKey key, TValue value)
+        {
+            if (key == null)
+            {
+                ThrowArgumentNullExceptionForKey();
+            }
+
+            if (_dictionaryStorage != null)
+            {
+                _dictionaryStorage.Add(key, value);
+                return;
+            }
+
+            EnsureCapacity(_count + 1);
+
+            if (_dictionaryStorage != null)
+            {
+                _dictionaryStorage.Add(key, value);
+                return;
+            }
+
+            if (ContainsKeyArray(key))
+            {
+                throw new ArgumentException($"An element with the key '{key}' already exists in the {nameof(AdaptiveCapacityDictionary<TKey, TValue>)}.", nameof(key));
+            }
+
+            _arrayStorage[_count] = new KeyValuePair<TKey, TValue>(key, value);
+            _count++;
+        }
+
+        /// <inheritdoc />
+        public void Clear()
+        {
+            if (_dictionaryStorage != null)
+            {
+                _dictionaryStorage.Clear();
+            }
+
+            if (_count == 0)
+            {
+                return;
+            }
+
+            Array.Clear(_arrayStorage, 0, _count);
+            _count = 0;
+            _arrayStorage = Array.Empty<KeyValuePair<TKey, TValue>>();
```

Comment:
Not sure we want to do this. Don't we want to keep the capacity the same but clear the storage. That's what this `TrimExess` is for. 

Don't we also need to Clear the dictionary

Reply:
We clear the dictionary above.

And I think you are correct here.